bundles/binding/org.openhab.binding.satel/src/main/java/org/openhab/binding/satel/internal/protocol/SatelModule.java

```diff
@@ -0,0 +1,455 @@
+
+	/*
+	 * Helper interface for connecting and disconnecting to specific module
+	 * type. Each module type should implement these methods to provide input
+	 * and output streams and way to disconnect from the module.
+	 */
+	protected interface CommunicationChannel {
+
+		InputStream getInputStream() throws IOException;
+
+		OutputStream getOutputStream() throws IOException;
+
+		void disconnect();
+	}
+
+	/*
+	 * Helper interface to handle communication timeouts.
+	 */
+	protected interface TimeoutTimer {
+
+		void start();
+
+		void stop();
+	}
+
+	/**
+	 * Creates new instance of the class.
+	 * 
+	 * @param timeout
+	 *            timeout value in milliseconds for connect/read/write
+	 *            operations
+	 */
+	public SatelModule(int timeout) {
+		this.integraType = IntegraType.UNKNOWN;
+		this.timeout = timeout;
+
+		addEventListener(this);
+		registerCommands();
+	}
+
+	/**
+	 * Returns type of Integra connected to the module.
+	 * 
+	 * @return Integra type
+	 */
+	public IntegraType getIntegraType() {
+		return this.integraType;
+	}
+
+	/**
+	 * Returns configured timeout value.
+	 * 
+	 * @return timeout value as milliseconds
+	 */
+	public int getTimeout() {
+		return this.timeout;
+	}
+
+	public boolean isConnected() {
+		return this.channel != null;
+	}
+
+	/**
+	 * Returns status of initialization.
+	 * 
+	 * @return <code>true</code> if module is properly initialized and ready for
+	 *         sending commands
+	 */
+	public boolean isInitialized() {
+		return this.integraType != IntegraType.UNKNOWN;
+	}
+
+	protected abstract CommunicationChannel connect();
+
+	/**
+	 * Starts communication. 
+	 */
+	public synchronized void open() {
+		this.communicationWatchdog = new CommunicationWatchdog();
+	}
+
+	/**
+	 * Stops communication by disconnecting from the module and stopping all background tasks. 
+	 */
+	public synchronized void close() {
+		if (this.communicationWatchdog != null) {
+			this.communicationWatchdog.close();
+			this.communicationWatchdog = null;
+		}
+	}
+
+	/**
+	 * Enqueues specified command in send queue if not already enqueued.
+	 * @param cmd command to enqueue
+	 * @return <code>true</code> if operation succeeded
+	 */
+	public boolean sendCommand(SatelMessage cmd) {
+		try {
+			if (this.sendQueue.contains(cmd)) {
+				logger.debug("Command already in the queue: {}", cmd);
+			} else {
+				this.sendQueue.put(cmd);
+				logger.trace("Command enqueued: {}", cmd);
+			}
+			return true;
+		} catch (InterruptedException e) {
+			return false;
+		}
+	}
+
+	/**
+	 * {@inheritDoc}
+	 */
+	@Override
+	public void incomingEvent(SatelEvent event) {
+		if (event instanceof IntegraVersionEvent) {
+			IntegraVersionEvent versionEvent = (IntegraVersionEvent) event;
+			this.integraType = IntegraType.valueOf(versionEvent.getType());
+			this.integraVersion = versionEvent.getVersion();
+			logger.info("Connection to {} initialized. Version: {}.", this.integraType.getName(), this.integraVersion);

```

Comment:
the statement doesn't seem to be consistent with the code since a VersionEvent is received not an InitialisationEvent (or alike). I'd prefer to either change the log statement or explain why receiving a VersionEvent leads to an initialised Connection